### PR TITLE
Fixing shadow variables in fem classes [shadow-fem-dev]

### PR DESCRIPTION
**Pull request by mlstowell**
This PR partially addresses issue #2701.

I tried to make minimal changes and keep the spirit of the original variable names. Hopefully, splitting these into small PRs will assist the original authors and/or reviewers in assessing the changes.

The compiler has no trouble determining which variable is intended but the choice is not always obvious to developers who may not be aware of one or the other of the possibilities.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#2739](https://github.com/mfem/mfem/pull/2739) | @mlstowell | @pazner | @pazner + @vladotomov | 12/28/21 | 2/18/22 | ⌛due 1/18/22 | |

### CODE REVIEW

File: fem/bilininteg.cpp
Comment (vladotomov) [inline fem/bilininteg.cpp:2668]: This seems counterproductive. Is `i` defined on the outside? Seems that the outside definition should change.

**Comment by pazner** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

File: fem/bilininteg.cpp
Comment (mlstowell) [reply fem/bilininteg.cpp:2668]: Yes, the loop over integration points uses `i`. However, this is a very common choice (~86% of our bilinear form integrators use `i` for this loop iterator) so I'd rather not change the outer definition in this case.

Comment (pazner) [reply fem/bilininteg.cpp:2668]: How about `vd` for the `vdim` iterator?

Comment (mlstowell) [reply fem/bilininteg.cpp:2668]: We need two iterators because these are accessing entries of a matrix-valued coefficient. I suppose we could use `vi` and `vj` or `di` and `dj` but are these really any better? The entire nested loop is only a ten lines long and these iterators are each used in two places. I think there's little change of confusion.

Comment (pazner) [reply fem/bilininteg.cpp:2668]: My personal preference would be `vd_i` and `vd_j` 🙂 

I agree there is not huge potential for confusion…

Comment (mlstowell) [reply fem/bilininteg.cpp:2668]: If @vladotomov agrees I'll be happy to make that change.

Comment (vladotomov) [reply fem/bilininteg.cpp:2668]: How about to make the loop over quadrature points `for (int q = 0 ...` and leave these `i` and `j`?
The outer index is used only at one place to do `ir->IntPoint(q)`

Comment (vladotomov) [reply fem/bilininteg.cpp:2668]: I see your comment about the 86%, but don't you agree `q` is better in this case?

Comment (mlstowell) [reply fem/bilininteg.cpp:2668]: I've used `q` in previous projects and it does seem natural to me. Unfortunately, `q` is more often used for coefficients in many MFEM classes.  The second most common iterator is `p` coming in at almost 11% (the only other one I see it `pint` which I don't like). It honestly doesn't matter much to me. My priorities were limiting the changes and/or following precedent. If @pazner is ok with using `q` (or `p`) either will be fine with me.

Comment (pazner) [reply fem/bilininteg.cpp:2668]: Because this discussion thread did not have enough suggestions, here's another one.

How about getting rid of the outer index altogether (since it's only used once, to get the integration point), and using a range-based for loop:
```c++
for (const IntegrationPoint &ip : *ir)
{
   […]
}
```

File: fem/restriction.cpp
Comment (vladotomov) [inline fem/restriction.cpp:1062]: I'd put `fe_trace`

File: fem/restriction.cpp
Comment (vladotomov) [inline fem/restriction.cpp:698]: There are many different `fe` names in this function, maybe @YohannDudouit should look at it too.

File: fem/pfespace.cpp
Comment (vladotomov) [inline fem/pfespace.cpp:753]: I'd use `ltdof_i`.

File: fem/fespace.cpp
Comment (vladotomov) [inline fem/fespace.cpp:1547]: I'd go with `fec_ref, mesh_ref`

File: fem/fespace.cpp
Comment (vladotomov) [inline fem/fespace.cpp:1652]: `subXt` unused.

File: fem/fespace.cpp
Comment (vladotomov) [inline fem/fespace.cpp:2120]: `this` not needed

File: fem/fespace.cpp
Comment (vladotomov) [inline fem/fespace.cpp:3559]: `NURBSext->GetNDof()` seems this be changed to `nurbs_ext`

I was starting to feel restless that I won't find an actual error.

**Review by vladotomov — CHANGES_REQUESTED** [review]

File: fem/pbilinearform.cpp
Comment (mlstowell) [reply fem/pbilinearform.cpp:320]: No, capital `X` and `Y` were member variables in the `ParBilinearForm` class but they were also used as arguments in at least three functions including `FormLinearSystem`. I changed the names of these member variables to `Xaux` and `Yaux` which lead to this change. I believe this is the only place these member variables are used.

File: fem/fespace.cpp
Comment (mlstowell) [reply fem/fespace.cpp:1652]: Good catch

File: fem/fespace.cpp
Comment (mlstowell) [reply fem/fespace.cpp:3559]: Good catch!

File: fem/estimators.hpp
Comment (mlstowell) [reply fem/estimators.hpp]: Fixed

File: fem/restriction.cpp
Comment (mlstowell) [reply fem/restriction.cpp]: Fixed

File: fem/fespace.cpp
Comment (mlstowell) [reply fem/fespace.cpp]: Fixed (this was a good suggestion)

File: fem/fespace.cpp
Comment (mlstowell) [reply fem/fespace.cpp]: Fixed

File: fem/pfespace.cpp
Comment (mlstowell) [reply fem/pfespace.cpp]: Fixed

**Review by vladotomov — APPROVED** [review]

File: fem/bilininteg.cpp
Comment (pazner) [inline fem/bilininteg.cpp:2950]: We can probably just use `dim` here like in the other integrators instead of `dim1`.

File: fem/linearform.cpp
Comment (pazner) [inline fem/linearform.cpp:172]: I think it's better to just move `int i` into the other for loops, `i` doesn't need to exist in a broader scope

File: fem/restriction.cpp
Comment (pazner) [inline fem/restriction.cpp:728]: Is `el` here the same as `tfe` above? @YohannDudouit

Comment (YohannDudouit) [reply fem/restriction.cpp:728]: Yes, but this code has also been completely refactored in #2195 , so I suspect that what these modifications are doing are mostly future merge conflicts.

File: fem/restriction.cpp
Comment (YohannDudouit) [reply fem/restriction.cpp:698]: This code has been completely refactored in #2195, I think it would be wiser to resolve these shadowing after the merge, no?

File: fem/restriction.cpp
Comment (pazner) [reply fem/restriction.cpp:728]: I agree, it would make sense to merge #2195 first, and then deal with the shadowing after that. I see that PR has two approvals already, is it waiting for @v-dobrev's final approval?

Comment (v-dobrev) [reply fem/restriction.cpp:728]: I'm not a reviewer and I have not looked at this PR -- if you want me to look at something specific, just let me know.

Comment (pazner) [reply fem/restriction.cpp:728]: @v-dobrev, I was referring to #2195 (DG + PA + NC AMR). That PR has two approvals, I believe your approval is the last outstanding one.

File: fem/linearform.cpp
Comment (mlstowell) [reply fem/linearform.cpp]: Fixed

File: fem/bilininteg.cpp
Comment (mlstowell) [reply fem/bilininteg.cpp]: Fixed

File: fem/restriction.cpp
Comment (v-dobrev) [reply fem/restriction.cpp:728]: Oh, yes, I will finish my review of that PR soon.

**Review by pazner — APPROVED** [review]
Approving this PR, but waiting until #2195 is merged to avoid conflicts.

**Comment by pazner** [comment]
> Approving this PR, but waiting until https://github.com/mfem/mfem/pull/2195 is merged to avoid conflicts.

@mlstowell, PR #2195 was just merged in master, when you get a chance could you update this branch and resolve the conflicts?

This should have an impact on `restriction.cpp` and `prestriction.cpp`.

**Comment by pazner** [comment]
Merged in `next` for testing...

**Pull request merged**
